colossalai/fx/passes/algorithms/ckpt_solver_pofo.py

```diff
@@ -0,0 +1,396 @@
+from typing import List, Tuple
+import copy
+import torch
+from torch.fx import GraphModule, Node
+from colossalai.fx.graph_module import ColoGraphModule
+from colossalai.fx.profiler import parameter_size
+import math
+from .linearize import linearize
+from .utils import *
+from colossalai.fx.passes.meta_info_prop import MetaInfoProp
+from colossalai.fx.codegen.activation_checkpoint_codegen import _find_nested_ckpt_regions
+from colossalai.fx.passes.algorithms.ckpt_solver_rotor import _construct_chain, _compute_table, _rec
+
+INF = float("inf")
+
+
+def _normalize_flops(chain: Chain, flops) -> Chain:
+    """
+    Normalize flops
+    """
+    for i in range(chain.length):
+        chain.fweight[i] /= flops
+        chain.bweight[i] /= flops
+
+    return chain
+
+
+class PofoTable:
+
+    def __init__(self, chain_length: int, mem_slots: int):
+        """Init pofo table
+        The pofo table contains two tables, opt and what, indicating values and
+        operations.
+
+        Args:
+            chain_length (int): chain length
+            mem_slots (int): number of memory slots
+        """
+
+        self.length = chain_length
+        self.mem_slots = mem_slots
+
+        # initializing tables
+        # the first bool indicates whether the input has bar
+        # opt table is for value, opt[True/False][i][A][(df, db)] = OCx(i, A, df, db)
+        # what table is for decision, what[True/False][i][A][(df, db)] = (is_enable, is_offload, index)
+        # where is_enable indicates whether we enable the gradient, is_offload indicates whether we
+        # offload the input, index indicates the end of F_\empty sequence if is_enable = False
+        self.opt = {
+            False: [[{} for _ in range(mem_slots + 1)] for _ in range(self.length + 1)],
+            True: [[{} for _ in range(mem_slots + 1)] for _ in range(self.length + 1)]
+        }
+        self.what = {
+            False: [[{} for _ in range(mem_slots + 1)] for _ in range(self.length + 1)],
+            True: [[{} for _ in range(mem_slots + 1)] for _ in range(self.length + 1)]
+        }
+
+    def _get_value(self, state, table, default):
+        i, act_size, df, db, input_has_bar = state
+        if act_size + df > self.mem_slots or act_size + db > self.mem_slots:
+            return default
+
+        try:
+            return table[input_has_bar][i][act_size][(df, db)]
+        except KeyError:
+            print(f"state not found {state}")
+
+    def get_opt(self, state):
+        return self._get_value(state, self.opt, INF)
+
+    def get_what(self, state):
+        return self._get_value(state, self.what, INF)
+
+    def set_value(self, state, opt, what):
+        i, act_size, df, db, input_has_bar = state
+        self.opt[input_has_bar][i][act_size][(df, db)] = opt
+        self.what[input_has_bar][i][act_size][(df, db)] = what
+
+
+class PofoSolver:
+
```

Comment:
You should include a docstring here. You can move your PR description here so that it will appear on our API  documentation.